### ISSUE TITLE
Wait For Condition Keyword should graciously stomach JavascriptExceptions

### ISSUE THREAD

**Issue by humbienri** [robotframework/SeleniumLibrary#1520]
Hello, please find the time to have a look-see down below  :)

# Prerequisites

- Searched for similar issues in open and closed tickets and cannot find a duplicate.
- Confirmed buggy behavior/issue exists in both versions 3.1.1 and 4.1.0 of SeleniumLibrary.
- I've already read through the User Guide and SeleniumLibrary's documentation without success.
- I am not using the Java Selenium2Library.
- No sensitive information being provided.

# For issues
## Steps to reproduce the issue

Create a simplified example which reproduces the problem. It is good if the example
can be run also by us, but this is not always mandatory.

## Error messages and additional information
##--loglevel trace stacktrace
```<kw name="Wait For Condition" library="SeleniumLibrary">
<doc>
Waits until ``condition`` is true or ``timeout`` expires.
</doc>
<arguments>
<arg>
return document.querySelector('div.acc-Document_HeaderInfoItemRow2 cmp-text-input input').value == '${ACCOUNTS_PAYABLE_INVOICE_NUMBER}';
</arg>
<arg>timeout=18s</arg>
</arguments>
<msg timestamp="20191209 17:24:39.957" level="TRACE">
Arguments: [ "return document.querySelector('div.acc-Document_HeaderInfoItemRow2 cmp-text-input input').value == 'QrLmgWXPIk';" | timeout='18s' ]
</msg>
<msg timestamp="20191209 17:24:39.957" level="DEBUG">
POST http://127.0.0.1:52264/session/c858e0d3adabe28dcb83003ceae0980e/execute/sync {"script": "return document.querySelector('div.acc-Document_HeaderInfoItemRow2 cmp-text-input input').value == 'QrLmgWXPIk';", "args": []}
</msg>
<msg timestamp="20191209 17:24:39.967" level="DEBUG">
http://127.0.0.1:52264 "POST /session/c858e0d3adabe28dcb83003ceae0980e/execute/sync HTTP/1.1" 500 1064
</msg>
<msg timestamp="20191209 17:24:39.968" level="DEBUG">Finished Request</msg>
<kw name="Capture Page Screenshot" library="SeleniumLibrary">...</kw>
<msg timestamp="20191209 17:24:40.603" level="FAIL">
JavascriptException: Message: javascript error: Cannot read property 'value' of null (Session info: chrome=78.0.3904.108)
</msg>
<msg timestamp="20191209 17:24:40.603" level="DEBUG">
Traceback (most recent call last): File "c:\python37\lib\site-packages\SeleniumLibrary\__init__.py", line 467, in run_keyword return DynamicCore.run_keyword(self, name, args, kwargs) File "c:\python37\lib\site-packages\SeleniumLibrary\base\robotlibcore.py", line 102, in run_keyword return self.keywords[name](*args, **kwargs) File "c:\python37\lib\site-packages\SeleniumLibrary\keywords\waiting.py", line 54, in wait_for_condition timeout, error File "c:\python37\lib\site-packages\SeleniumLibrary\keywords\waiting.py", line 265, in _wait_until self._wait_until_worker(condition, timeout, error) File "c:\python37\lib\site-packages\SeleniumLibrary\keywords\waiting.py", line 272, in _wait_until_worker if condition(): File "c:\python37\lib\site-packages\SeleniumLibrary\keywords\waiting.py", line 52, in <lambda> lambda: self.driver.execute_script(condition) is True, File "c:\python37\lib\site-packages\selenium\webdriver\remote\webdriver.py", line 636, in execute_script 'args': converted_args})['value'] File "c:\python37\lib\site-packages\SeleniumLibrary\keywords\browsermanagement.py", line 660, in execute result = self._base_execute(driver_command, params) File "c:\python37\lib\site-packages\selenium\webdriver\remote\webdriver.py", line 321, in execute self.error_handler.check_response(response) File "c:\python37\lib\site-packages\selenium\webdriver\remote\errorhandler.py", line 242, in check_response raise exception_class(message, screen, stacktrace)
</msg>
<status status="FAIL" starttime="20191209 17:24:39.957" endtime="20191209 17:24:40.603"/>
</kw>
```

## Expected behavior and actual behavior

I would expect the default behaviour would  be to silently ignore the javascript error. In my particular instance, the web browser is transitioning from one page to another and thus, depending on a slow system-under-test, the web element of interest may NOT quite be on page just yet.

I expected that specifying a `timeout` on the `Wait For Condition` KW would have alleviated the problem but the KW still fails out pretty fast and is evident that the timeout isn't being honored.

`Wait For Condition` keyword should be able to graciously handle a JavascriptException without erroring and immediately failing the test. This could simply be, ignore the specific exception type and not error out until the timeout has expired.

After some time, effort, and sleuthing I think the offending method might be in the `waiting.py` module, in the `_wait_until_worker` where a `StaleElementReferenceException` is already being handled & ignored.

## Environment

Browser: chrome=78.0.3904.108
Browser driver: Windows 10 Pro
Operating System: Windows 10 Pro
Libraries
- Robot Framework: 3.1.1
- Selenium: 3.141.0
- SeleniumLibrary: 3.3.1
- Interpreter: python 3.7.1

**Comment by aaltat** [comment]
We don't usually protect users from this kind of errors. In the past it did mean that something really bad did happen and users should know about it. But the way web applications are being done has changed the reasoning of the past. 

We did have a discussion should the wait keywords suppress the `StaleElementReferenceException` or not. Both ways have good reasons to write the implementation in their way. But in the end the usability of the library won and implementation was changed to suppress `StaleElementReferenceException`. Your request feels like a similar need and we need to discuss it with larger community. I will post your request in the slack and let's see what happens. If you want your side to win, writing few examples would be good thing.

**Comment by humbienri** [comment]
Hello Tatu!

I didn't realize this had already been debated back 'n forth.  I don't know exactly what you mean by a few examples?  I don't have the skills to write up HTML pages or a small web app. I am not a web developer but a software tester.  Or maybe I am misunderstanding you.

Also as to the `Wait...` keywords needing to or not, swallow up Stale Element Reference Exceptions, to me it seems kind of logical for them to swallow and suppress those Exceptions.  I think the point of the `Wait...` KWs is to provide flexibility for synchronization between the automated test and the UI for the SUT no? As such, they should provide some "elasticity" and wiggle room.

The workaround I have instituted for now is to have both a `Wait for Element is Visible` followed by a `Wait For Condition` and they both use the same locator but the 2nd one tries to get its value via the passed in Javascript.  It works but I think it is ugly and a hack.  I actually accidentally stumbled across this problem, the topic of this issue, by accident when I commented out the `Visible` line.

I've not looked closely at the source code for the `Wait...Visible` Keyword as to why it is that it does not trip up with the Javascript Null object like the `Wait For Condition` does.

Thank you for your time sir :)

**Comment by aaltat** [comment]
I understand you point, but I can also see it from the other side that for some users this type of errors should not be suppressed. Because there is a clear need for the both ways, it is decision which is hard to make. I  am currently leaning towards keeping the current implementation, but because `Wait For Condition` is one of the rare keywords which I have not used in work life, I do not feel strongly on the matter. Therefore I did ask community for feedback, in the [slack](https://robotframework-slack-invite.herokuapp.com/),  but the question did not get much of traction.  So if you feel strongly on the matter, please join the slack community or post the user group mailing list and try to get support for change request. 

Another option is to catch the JavaScript error in the JavaScript side and return `null`, example as in this [StackOverflow](https://stackoverflow.com/questions/7120290/how-to-suppress-all-javascript-runtime-errors)

**Comment by humbienri** [comment]
I THINK I understand your point or your hesitation perhaps? I see that all of the `Wait...` Keywords use `_wait_until` and `_wait_until_worker` python methods in `waiting.py` so adding the suppression of a `JavascriptException` to that method would ripple out to other/all Wait KWs?  Is that the fear?

How long has this KW, `Wait For Condition` been around that you yourself have also not used it in real life?  I wonder if others have also not used it extensively or perhaps not in a scenario of transitioning from one page to another page. Perhaps?

Thanks Tatu.

**Comment by aaltat** [comment]
Technically this is not difficult to do and should not complicated the code base too much. But what I fear is the lack of data. I personally have not used this keyword in work, so I do not have a good gut feeling about it. It is just something that I have not needed and this is not the only area in the SeleniumLibrary which I have not used in work. 

You are the first one to request this type of change for this keyword, I do not recall seeing this type of request in the Slack or in the user group mailing list. I did ask community opinion, in the Slack few times, but it did got only one response (which as against the change). Therefore the data from the community is limited. Also the current implementation has purpose and I can imagine scenarios where the current implementation is correct. 

Also the writing the code is small part of the cost, it is the maintenance cost that will kill. Example PR is only done once, but project needs support. Therefore maintainers usually think support in years (example in core Python project they think in decades) and at least I think that I need to support every line of the code in 5 years time span (in one form or other). Also there is a cost for users, (based on the user group data and random questions in Slack, keyword is used) if keywords changes, users need to change they test data when they update the library if the new functionality does not suite to their needs. But I do not have visibility on how large user base is for this keyword and how much it would cost user to do the change. But there is a cost for the users, that I am sure of.  

The biggest obstacle for this feature is that it backwards incompatible. Therefore making the change would need community to get interested and support the change. But I am not seeing that at the moment. For the `Wait...` keywords we did discuss the matter, on different kind of forums about an year, because it was very big change in the fundamental level (The code [change](https://github.com/robotframework/SeleniumLibrary/commit/8120ef44eb7086275280d8c9f1a91d60938b93f6) it self is trivial). Also there was people against the change and supporting the change. Big majority of community did support the change and therefore in the end it was made. And I was against the change, because in my opinion it leaves gaps in testing and solves the right problem in an incorrect way. But because the majority of the community needed and wanted the change, eventually I did make it. So if you want this change to happen try to rally community behind your idea and wanting the change.
